### PR TITLE
ADS GetObjectNames method - Add params to get sorted or hidden object names

### LINKED ISSUE #16500: ADS should return sorted object names
See discussion in #16496.

At present, the ADS returns workspace names (via `getObjectNames()`) as an unordered set, which then has to be sorted. Sorting occurs in several separate places:
- WorkspaceProperty
- WorkspaceSelector
- "Add workspace" dialog in Multi dataset fitting interface
(and maybe more?)

It would be nicer if `getObjectNames()` returned an ordered collection (already sorted) instead.

### PR DESCRIPTION
**Pull request by DavidFair**
Description of work.
Previously the ADS `getObjectNames` method returned an unordered set of names. Many of the callers of this method subsequently sorted these names. 

The method has been rewritten to allow users to specify if they want the names sorted whilst the `getObjectNamesInclHidden()` has been merged in as an additional hidden parameter. 
The default parameters still return a unsorted list with hidden items included if set in the configuration as per the original behaviour.

All callers of this method have been checked and any which subsequently sorted the values have been refactored.

**To test:**

All units tests should pass. 
WorkspaceSelector and the Add workspace dialogue in multi dataset fitting should return sorted lists still.

Fixes #16500 .

Doesn't need to be added to release notes.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by raquelalvarezbanos** [comment]
Looks good to me: `getObjectNames()` is now able to return unordered/ordered vector of workspace names (it is also possible to specify whether or not hidden workspaces should be included). `WorkspaceProperty`, `ADSValidator`, `MDFAddWorkspaceDialog` and `WorkspaceSelector` have been updated accordingly. There's also a new unit test testing the changes.

:shipit:

File: Framework/Kernel/inc/MantidKernel/DataService.h [hunk 1]
```diff
@@ -19,6 +19,15 @@
...
 namespace Mantid {
 namespace Kernel {
+
+/// Flag for whether to sort items before returning
+enum class DataServiceSort { Sorted, Unsorted };
+/**
+ * Flag for whether to include hidden items when returning,
+ * Auto queries the class to determine this behavior
+ */
+enum class DataServiceHidden { Auto, Include, Exclude };
```
Comment (FedeMPouzols) [inline Framework/Kernel/inc/MantidKernel/DataService.h]: When you have an enum that is specific to a class, an alternative is to declare that enum inside the class. That's a way of limiting its scope, similarly as you would do with namespaces. So you would have something like `Mantid::Kernel::DataService::SortOption` or `Mantid::Kernel::DataSerive::HiddenMode`.
I don't necessarily have a strong preference, but maybe this is a very clear-cut case? Notice the names of the enums are `DataService...`.

**Comment by FedeMPouzols** [comment]
I think that the changes in this PR are good and tidy up the code, but I'm very confused by the traditional use of `map` and `unordered_map` in the `DataService`. The underlying structure is an ordered `std::map`  (typedef `svcmap`) but `getObjectNames()`  used to take the non-hidden objects from the map and insert them in an `unordered_map`, which would then need sorting.

I wouldn't be very concerned with performance as I think we normally have smallish collections of workspaces or instruments, but has there been any thought in the past as to whether the `DataService`s should keep its elements sorted, and what data structure it is supposed to be? @martyngigg, @NickDraper 

I'm wondering if, related to [this maintenance issue](https://github.com/mantidproject/mantid/issues/15370), the map underlying the analysis service should be an `unordered_map`.

Also, when I see that we now have the `Sorted` and `Unsorted` alternatives in getObjectNames(), I wonder if anyone willingly uses the `Unsorted` option, other than tests.

**Comment by martyngigg** [comment]
@FedeMPouzols I think you are right to question the conflicted uses of these types. In C++03 days without tr1 then `map`  was the only `std` structure that could be used. These days I think `unordered_map` is the way to go in most situations as it's actually really a hash table and you're getting what you really want which is just a way to associated a value with a key.

I agree that part of the referenced maintenance issue should convert the underlying storage type from `map` to `unordered_map`.

On the point about the `Unsorted` option - I don't think it will be used much either but as it was simple to do we thought why not include it.

File: Framework/Kernel/inc/MantidKernel/DataService.h (hunk 1 above)
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DataService.h]: @FedeMPouzols David chatted to me about this and usually I would agree with you. What we didn't like much was that the templated type would need to be included when calling the enum e.g.  `DataService<Workspace_sptr>::Sorted`, so this seemed like a reasonable compromise. What do you think?

Comment (FedeMPouzols) [inline Framework/Kernel/inc/MantidKernel/DataService.h]: Ah, I hadn't considered that here there's the additional complexity of the template. That makes sense as a good compromise.

Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DataService.h]: Thanks. I'll get this merged now.

**Comment by FedeMPouzols** [comment]
Thanks @martyngigg, that makes sense. I've added a comment in the maintenance issue.

**Pull request merged**